pkg/operator/controller/dns/controller.go

```diff
@@ -191,6 +193,53 @@ func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 	return result, nil
 }
 
+// createDNSProviderIfNeeded creates a new DNS provider if none has yet been
+// created or if the infrastructure platform status or cloud credentials have
+// changed since the current provider was created.  After creating a new
+// provider, createDNSProviderIfNeeded updates the reconciler state
+// with the new provider and current platform status and cloud credentials.
+func (r *reconciler) createDNSProviderIfNeeded(dnsConfig *configv1.DNS) error {
+	var needUpdate bool
+
+	infraConfig := &configv1.Infrastructure{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, infraConfig); err != nil {
+		return fmt.Errorf("failed to get infrastructure 'config': %v", err)
+	}
+
+	platformStatus, err := getPlatformStatus(r.client, infraConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get platform status: %v", err)
+	}
+
+	creds := &corev1.Secret{}
+	switch platformStatus.Type {
+	case configv1.AWSPlatformType, configv1.AzurePlatformType, configv1.GCPPlatformType:
+		name := types.NamespacedName{Namespace: r.Config.Namespace, Name: cloudCredentialsSecretName}
+		if err := r.cache.Get(context.TODO(), name, creds); err != nil {
```

Comment:
I noticed before these changes `creds` was being populated by a call to `r.client.Get` instead of `r.cache.Get`. Just wanted to verify that this is intentional and won't cause any defects.

Reply:
Yeah, it is intentional.  Want to avoid incurring the API call on every reconciliation, and it may avoid cache consistency issues between the watch on the infrastructure resource that triggers the reconciliation and the reconciliation logic here that reads the infrastructure object.  It might have made sense to make this change in #417, which moved this logic from the operator startup (where we don't yet have the cache) to the reconciliation logic (where we do).  It definitely makes sense to use the cache if we're using the value on every reconciliation.